core/src/main/java/io/grpc/Context.java

```diff
@@ -0,0 +1,496 @@
+package io.grpc;
+
+import com.google.common.base.Preconditions;
+import com.google.common.util.concurrent.MoreExecutors;
+
+import java.io.Closeable;
+import java.io.IOException;
+import java.util.ArrayDeque;
+import java.util.ArrayList;
+import java.util.concurrent.Callable;
+import java.util.concurrent.Executor;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.atomic.AtomicIntegerFieldUpdater;
+
+import javax.annotation.Nullable;
+
+
+/**
+ * --Add warning about using this too liberally--
+ * --Add warning about using this with containers that unload Classloaders--
+ * -- Add note that in the case of cascading closure that listeners to the parent are always
+ * notified before listeners on the children.
+ *
+ */
+public class Context {
```

Comment:
I think one issue is that the functionality here isn't really specific to GRPC, at some point it could easily move into it's own thing. Given that and the fact that the package would only have one or two classes in it I'm not sure it makes much of a difference either way.